[LLDB] Update DIL to pass current 'frame var' tests.

cmtice opened the pull request:
As a preliminary to making DIL the default implementation for 'frame var', ran check-lldb forcing 'frame var' to always use DIL, and discovered a few failing tests. This fixes most of them. The only two remaining failing tests (once the smart pointer PR is committed) are TestVarPath.py, which fails a test case using a negative array subscript, as DIL does not yet parse negative numbers; and TestDAP_evaluate.py, which now passes a test case that the test says should fail (still investigating this).

Changes in this PR:
 - Sets correct VariableSP, as well as returning ValueObjectSP (needed for several watchpoint tests).
 - Updates error messages, when looking up members, to match what the rest of LLDB expects. Also update appropriate DIL tests to expect the updated error messages.
 - Updates DIL parser to	look for and accept "(anonymous namespace)::" at the front of a variable name.

cmtice committed:
[LLDB] Update DIL to pass current 'frame var' tests.

As a preliminary to making DIL the default implementation for 'frame var',
ran check-lldb forcing 'frame var' to always use DIL, and discovered a few
failing tests. This fixes most of them. The only two remaining failing tests
(once the smart pointer PR is committed) are TestVarPath.py, which fails a
test case using a negative array subscript, as DIL does not yet parse negative
numbers; and TestDAP_evaluate.py, which now passes a test case that the test says should fail (still investigating this).

Changes in this PR:
 - Sets correct	VariableSP, as well as returning ValueObjectSP (needed for
   several watchpoint tests).
 - Update error	messages, when looking up members, to match what the rest of
   LLDB expects. Also update appropriate DIL tests to expect the updated error
   messages.
 - Update DIL parser to	look for and accept "(anonymous namespace)::" at the
   front of a variable name.

cmtice commented:
Question: Should we update DIL to parse/handle negative integers before making the switch? (e.g. to pass the TestVarPath test)

cmtice committed:
Fix clang-format issues.

cmtice committed:
Fix more clang-format issues.

cmtice committed:
Fix clang-format issue.

labath commented on lldb/source/ValueObject/DILParser.cpp:
```diff
@@ -178,11 +178,40 @@ ASTNodeUP DILParser::ParsePrimaryExpression() {
+      // Save the source location for the diagnostics message.
+      uint32_t loc = CurToken().GetLocation();
+      m_dil_lexer.Advance();
+      assert(
+          (CurToken().Is(Token::identifier) || CurToken().Is(Token::l_paren)) &&
+          "Expected an identifier or anonymous namespeace, but not found.");
```
An assert is fine for an internal invariant, but this looks like its checking the next token in the stream. That's can't be right. Or am I misunderstanding something?

labath replied:
No -- I don't think I am -- I just made lldb crash with this assertion.

labath replied:
```
(lldb) frame variable 'lldb::(anonymous namespace)::a'
lldb: /home/pavelo/ll/mono/lldb/source/ValueObject/DILParser.cpp:235: std::string lldb_private::dil::DILP
arser::ParseNestedNameSpecifier(): Assertion `(CurToken().Is(Token::identifier) || CurToken().Is(Token::l
_paren)) && "Expected an identifier or anonymous namespace, but not found."' failed.
```

labath commented on lldb/source/ValueObject/DILParser.cpp:
```diff
@@ -178,11 +178,40 @@ ASTNodeUP DILParser::ParsePrimaryExpression() {
+        m_dil_lexer.LookAhead(4).Is(Token::coloncolon)) {
+      m_dil_lexer.Advance(4);
+
+      std::string identifier = "(anonymous namespace)";
```
Instead of basically inlining a part of ParseIdExpression, could we find a way to delegate to it? What would happen if we changed the check on line 172 to say "if next token is :: or identifier *OR* next 4 tokens are '(anonymous namespace)'" ?

labath reviewed (commented):
If the test you found the anonymous namespace issue it is explicitly testing the handling of anonymous namespaces, then this is fine. However, if it's doing that as a side-effect of testing something else, then it may be nice to have an explicit test case for it. (Particularly as the anonymous namespace thing can appear in other positions as well (`ns1::(anonymous namespace)::foo`, `(anonymous namespace)::ns::(anonymous namespace)::foo`, etc.)

cmtice committed:
Merge remote-tracking branch 'origin/main' into DIL-fix-tests

cmtice committed:
Address reviewer comments:

- Replace assert statements with appropriate error checks & handling.
- Consolidate code checking for 'anonymous namespace'.
- Fix various small issues; add checks for 'anonymous namespaces' in
  multiple different positions.

cmtice committed:
Merge remote-tracking branch 'origin/main' into DIL-fix-tests

cmtice committed:
Fix bug introduced in previous commit.

cmtice commented:
I think I have addressed all your concerns. Please take another look; thanks!

labath commented on lldb/source/ValueObject/DILParser.cpp:
```diff
@@ -232,16 +248,16 @@ std::string DILParser::ParseNestedNameSpecifier() {
+      m_dil_lexer.Advance();
+      if (!CurToken().Is(Token::identifier) && !CurToken().Is(Token::l_paren)) {
+        BailOut(
+            "Expected an identifier or anonymous namespeace, but not found.",
```
namespace

cmtice replied:
Done.

labath reviewed (commented):
It's better, but it still looks like there's an inlined copy of ParseIdExpression inside ParsePrimaryExpression. I'd be nice to avoid that as it's easy to forget the inlined version when updating ParseIdExpression (and the problem would show up on variables inside anonymous namespaces).

Maybe that's impossible, but if so, I'd like to know why.

cmtice committed:
Address reviewer comment:

Rearrange code slightyl to ParsePrimaryExpression doesn't duplicate
code in ParseIdExpression. Also fix type & return-after-else.

labath reviewed (approved):
Thanks.
